crates/sui-graphql-rpc/src/types/balance.rs

```diff
@@ -160,3 +205,54 @@ impl TryFrom<StoredBalance> for Balance {
         })
     }
 }
+
+pub(crate) fn build_candidates(snapshot_objs: RawQuery, history_objs: RawQuery) -> RawQuery {
+    let candidates = query!(
+        r#"SELECT DISTINCT ON (object_id) * FROM (
+        ({})
+        UNION
+        ({})
+    ) o"#,
+        snapshot_objs,
+        history_objs
+    );
+
+    candidates
+        .order_by("object_id")
+        .order_by("object_version DESC")
+}
+
+pub(crate) fn build_newer(lhs: i64, rhs: i64) -> RawQuery {
+    let newer = query!(r#"SELECT object_id, object_version FROM objects_history"#);
+    filter!(
+        newer,
+        format!(r#"checkpoint_sequence_number BETWEEN {} AND {}"#, lhs, rhs)
+    )
+}
+
+pub(crate) fn build_join(candidates: RawQuery, newer: RawQuery) -> RawQuery {
+    let final_ = query!(
+        r#"
+    SELECT CAST(SUM(candidates.coin_balance) AS TEXT) as balance, COUNT(*) as count, candidates.coin_type as coin_type
+        FROM ({}) candidates
+        LEFT JOIN ({}) newer
+        ON (
+            candidates.object_id = newer.object_id
+            AND candidates.object_version < newer.object_version
+        )"#,
```

Comment:
nit: whitespace.

```suggestion
        r#"SELECT 
            CAST(SUM(candidates.coin_balance) AS TEXT) as balance, 
            COUNT(*) as count, 
            candidates.coin_type as coin_type
        FROM ({}) candidates
        LEFT JOIN ({}) newer
        ON (
            candidates.object_id = newer.object_id
            AND candidates.object_version < newer.object_version
        )"#,
```

Also on my earlier question of whether you could avoid mentioing `candidates.` in the cursor query, I think the answer is yes, if you also join on `candidates.coin_type = newer.coin_type` which should be true because coin types are not going to change between different versions of the same object (we hope).

Reply:
I also have a little bit of a hard time figuring out why the query building is split up into multiple functions. I can see the benefit of doing that for the innermost `filter` application, but otherwise it's making things quite hard to follow, because I'm jumping around.

Maybe I'm missing a place where these things are being re-used, in which case feel free to ignore me, but if it's purely to split things up then I would recommend opting to have them in a single function, but use comments to split that up into recognisable sections.

Reply:
That's fair, I'll put it back into one function.

Reply:
And unfortunately, even with the additional join, postgres will still complain that just `coin_type` is ambiguous. Just tried it manually :/